### DISCUSSION TITLE
How to actually add physics calculations into an armi run?

### DISCUSSION THREAD

**Discussion by oscars47** [terrapower/armi#1687] [Q&A] [answered]
Hi Armi community,

I am new to nuclear physics and for a school project I am trying to simulate how the output power changes as a function of the fuel rod structure, but I am having difficulty getting a single simulation to actually run.

<img width="964" alt="Screenshot 2024-04-14 at 11 03 17 PM" src="https://github.com/terrapower/armi/assets/106777951/73bf38ee-dd9d-48aa-81a8-43fc5d7cbbae">

I am seeing no actual time evolution (0.00) throughout the run. I am using the files from the tutorial (anl-afci-177): https://terrapower.github.io/armi/tutorials/data_model.html. 

The python code I am running for the simulation is:
```
import armi
if not armi.isConfigured():
    armi.configure(armi.apps.App())

o=armi.init(fName="anl-afci-177.yaml")
o.operate()
```

Any advice would be greatly appreciated! I just need to see a complete working simulation. I have explored https://github.com/terrapower/fftf-isothermal-model, but this also does not explain how to interface with the physics plugin. I tried following this documentation (https://terrapower.github.io/dragon-plugin/), but it is outdated (there is no module ```terrapower```. I also tried looking at https://github.com/terrapower/armicontrib-dif3d, but in trying to run dif3ddemo/app.py, I get the error ``ModuleNotFoundError: No module named 'dif3ddemo'``. 

Thanks in advance for all your help!
Oscar

**Comment by john-science** [comment]
So, ARMI is open source, so we can not include physics calculations. We are just a framework.

The idea is that you have to provide your own actual physics calculations, in the form of an ARMI Plugin, like the ones you mention above. You would have to get [DIF3D](https://www.anl.gov/nse/dif3d) .

**Comment by jakehader** [comment, marked as answer]
Hi @oscars47, welcome! And thanks for the question! As @john-science said, the ARMI framework is developed and open sourced to provide others with a base set of infrastructure to build what are called "Applications" or "apps" on top of it to support the design and analysis, including sensitivity studies like you're interested in. Our teams both develop this framework and build our "apps" internally for reactor design calculations. 

We have taken some time to build example plugins to some codes, such as DIF3D, OPENMC (still a work in progress), and DRAGON to show examples of how this may be done but I would not characterize these examples as being fully generalized or without dependencies. In most cases, the actual software that is running behind this scenes are codes like DIF3D (neutron diffusion/transport solver from Argonne National Lab), MCNP (LANL), etc. There are limited codes available in Open Source for nuclear engineering analysis out of the box, but I might suggest looking into OpenMC (https://docs.openmc.org/en/stable/). Installation instructions here -https://docs.openmc.org/en/stable/quickinstall.html

OpenMC is a Monte Carlo based code, similar to MCNP that supports generalized geometry and material inputs. Depending on the size of the model and the tallies of interest you might need some larger computing power, but this might be your best bet to support a school project. @ntouran created an example plugin for this code here - https://github.com/terrapower/armicontrib-openmc, which could be expanded on to integrate with the FFTF and AFCI example problems, but this would likely take some time to understand the work you are doing, the timeline, etc. 

There are probably several caveats to what I am saying, but I am happy to set up a call with you if you'd like to discuss further! Please send me an email at jhader@terrapower.com and we can set up a 30 minute - 1 hour call. I may be able to point you in a direction to help your project as well as give some pointers to ARMI / OpenMC!

I am free this week so feel free to send me a message with some times you might be available and we can coordinate. Although we might not have what you need readily available, we hope to help and that you stick around to see the progress of the framework and / or support improvements on the framework or Open Source plugins. 

The minimum viable product for some studies may not be that far off with OpenMC and the plugin and I look forward to a discussion if you have the time / interest!

- Jake
